File: src/lib/scheduler/genetic-optimizer.ts
/**
 * =============================================================================
 * WARSTWA 3: GENETIC OPTIMIZER (ALGORYTM GENETYCZNY)
 * =============================================================================
 *
 * Finalne doszlifowanie grafiku metodą ewolucyjną.
 * Optymalizuje soft constraints: preferencje, bloki zmian, komfort pracowników.
 *
 * Operatory genetyczne:
 * - Selekcja turniejowa
 * - Krzyżowanie jednopunktowe
 * - Mutacje: swap, move, change template
 */

import {
    type SchedulerInput,
    type GeneratedShift,
    type EmployeeWithData,
} from "./types";
import { getDayOfWeek } from "./scheduler-utils";
import { logger } from "@/lib/utils/logger";
import { quickFitness } from "./evaluator";
import { checkHardConstraintsForShift } from "./validation";
import { getAvailableTemplatesForEmployee } from "./scheduler-utils";
import { DAY_KEYS } from "@/lib/constants/days";

// =============================================================================
// KONFIGURACJA
// =============================================================================

export interface GeneticConfig {
    populationSize: number;
    generations: number;
    mutationRate: number;
    crossoverRate: number;
    eliteCount: number;
    tournamentSize: number;
    timeoutMs: number;
}

export const DEFAULT_GENETIC_CONFIG: GeneticConfig = {
    populationSize: 30,
    generations: 100,
    mutationRate: 0.2,
    crossoverRate: 0.7,
    eliteCount: 2,
    tournamentSize: 3,
    timeoutMs: 5000,
};

export const FAST_GENETIC_CONFIG: GeneticConfig = {
    populationSize: 15,
    generations: 30,
    mutationRate: 0.25,
    crossoverRate: 0.6,
    eliteCount: 1,
    tournamentSize: 2,
    timeoutMs: 2000,
};

// =============================================================================
// TYPY
// =============================================================================

interface Individual {
    shifts: GeneratedShift[];
    fitness: number;
}

// =============================================================================
// GŁÓWNA KLASA
// =============================================================================

export class GeneticOptimizer {
    private config: GeneticConfig;
    private input: SchedulerInput;
    private allWorkingDays: string[];
    private weekendDays: Set<string>;

    constructor(input: SchedulerInput, config: Partial<GeneticConfig> = {}) {
        this.config = { ...DEFAULT_GENETIC_CONFIG, ...config };
        this.input = input;

        this.allWorkingDays = [
            ...input.workDays,
            ...input.saturdayDays,
            ...input.tradingSundays,
        ].sort();

        this.weekendDays = new Set([
            ...input.saturdayDays,
            ...input.tradingSundays,
        ]);
    }

    /**
     * Optymalizuje grafik algorytmem genetycznym
     */
    optimize(shifts: GeneratedShift[]): GeneratedShift[] {
        const startTime = Date.now();

        logger.log("\n🧬 === GENETIC OPTIMIZER - START ===");
        logger.log(
            `Populacja: ${this.config.populationSize}, Generacje: ${this.config.generations}`,
        );

        const initialFitness = quickFitness(shifts, this.input);
        logger.log(`Fitness początkowy: ${initialFitness}`);

        // Inicjalizuj populację
        let population = this.initializePopulation(shifts);

        let bestIndividual = population[0];
        let generationsWithoutImprovement = 0;

        // Główna pętla ewolucji
        for (let gen = 0; gen < this.config.generations; gen++) {
            // Timeout
            if (Date.now() - startTime > this.config.timeoutMs) {
                logger.log(`⏱️ Timeout po ${gen} generacjach`);
                break;
            }

            // Sortuj po fitness
            population.sort((a, b) => b.fitness - a.fitness);

            // Nowa generacja
            const newPopulation: Individual[] = [];

            // Elitaryzm
            for (let i = 0; i < this.config.eliteCount; i++) {
                newPopulation.push(this.clone(population[i]));
            }

            // Generuj potomków
            while (newPopulation.length < this.config.populationSize) {
                const parent1 = this.tournamentSelect(population);
                const parent2 = this.tournamentSelect(population);

                let child: Individual;
                if (Math.random() < this.config.crossoverRate) {
                    child = this.crossover(parent1, parent2);
                } else {
                    child = this.clone(parent1);
                }

                if (Math.random() < this.config.mutationRate) {
                    this.mutate(child);
                }

                child.fitness = quickFitness(child.shifts, this.input);
                newPopulation.push(child);
            }

            population = newPopulation;

            // Sprawdź najlepszego
            const currentBest = population.reduce((a, b) =>
                a.fitness > b.fitness ? a : b,
            );

            if (currentBest.fitness > bestIndividual.fitness) {
                bestIndividual = currentBest;
                generationsWithoutImprovement = 0;
            } else {
                generationsWithoutImprovement++;
            }

            // Early stopping
            if (generationsWithoutImprovement > 30) {
                logger.log(`🛑 Early stopping po ${gen} generacjach`);
                break;
            }
        }

        const elapsed = Date.now() - startTime;
        const improvement = bestIndividual.fitness - initialFitness;

        logger.log(`\n✅ GENETIC OPTYMALIZACJA ZAKOŃCZONA`);
        logger.log(`Czas: ${elapsed}ms`);
        logger.log(
            `Fitness: ${initialFitness} → ${bestIndividual.fitness} (${
                improvement > 0 ? "+" : ""
            }${improvement})`,
        );
        logger.log("🧬 === GENETIC OPTIMIZER - KONIEC ===\n");

        // Ostateczna deduplikacja przed zwróceniem
        this.fixConflicts(bestIndividual.shifts);

        return bestIndividual.shifts;
    }

    // =========================================================================
    // INICJALIZACJA
    // =========================================================================

    private initializePopulation(
        initialShifts: GeneratedShift[],
    ): Individual[] {
        const population: Individual[] = [];

        // Oryginalny osobnik
        population.push({
            shifts: [...initialShifts],
            fitness: quickFitness(initialShifts, this.input),
        });

        // Mutacje oryginału
        for (let i = 1; i < this.config.populationSize; i++) {
            const mutant: Individual = {
                shifts: [...initialShifts],
                fitness: 0,
            };

            // Losowa liczba mutacji
            const mutationCount = Math.floor(Math.random() * 3) + 1;
            for (let m = 0; m < mutationCount; m++) {
                this.mutate(mutant);
            }

            mutant.fitness = quickFitness(mutant.shifts, this.input);
            population.push(mutant);
        }

        return population;
    }

    // =========================================================================
    // OPERATORY GENETYCZNE
    // =========================================================================

    private tournamentSelect(population: Individual[]): Individual {
        let best: Individual | null = null;

        for (let i = 0; i < this.config.tournamentSize; i++) {
            const idx = Math.floor(Math.random() * population.length);
            const candidate = population[idx];

            if (!best || candidate.fitness > best.fitness) {
                best = candidate;
            }
        }

        return best!;
    }

    private crossover(parent1: Individual, parent2: Individual): Individual {
        // Jednopunktowe krzyżowanie po dniach
        const days1 = this.groupByDay(parent1.shifts);
        const days2 = this.groupByDay(parent2.shifts);

        const allDays = [...new Set([...days1.keys(), ...days2.keys()])].sort();
        const crossoverPoint = Math.floor(Math.random() * allDays.length);

        const childShifts: GeneratedShift[] = [];

        for (let i = 0; i < allDays.length; i++) {
            const day = allDays[i];
            const source = i < crossoverPoint ? days1 : days2;
            const dayShifts = source.get(day) || [];
            childShifts.push(...dayShifts);
        }

        // Napraw konflikty (duplikaty pracownik-dzień)
        this.fixConflicts(childShifts);

        return {
            shifts: childShifts,
            fitness: 0,
        };
    }

    private mutate(individual: Individual): void {
        const operation = Math.random();

        if (operation < 0.4) {
            // Swap: zamień pracowników między zmianami
            this.mutateSwap(individual);
        } else if (operation < 0.7) {
            // Move: przenieś zmianę na inny dzień
            this.mutateMove(individual);
        } else {
            // Change template: zmień szablon zmiany
            this.mutateTemplate(individual);
        }
    }

    private mutateSwap(individual: Individual): void {
        if (individual.shifts.length < 2) return;

        const idx1 = Math.floor(Math.random() * individual.shifts.length);
        let idx2 = Math.floor(Math.random() * individual.shifts.length);
        while (idx2 === idx1) {
            idx2 = Math.floor(Math.random() * individual.shifts.length);
        }

        const shift1 = individual.shifts[idx1];
        const shift2 = individual.shifts[idx2];

        // Sprawdź czy zamiana jest możliwa
        if (shift1.date === shift2.date) return; // Ten sam dzień
        if (shift1.employee_id === shift2.employee_id) return; // Ten sam pracownik

        const emp1 = this.input.employees.find(
            (e) => e.id === shift1.employee_id,
        );
        const emp2 = this.input.employees.find(
            (e) => e.id === shift2.employee_id,
        );
        if (!emp1 || !emp2) return;

        // Sprawdź czy pracownicy mogą pracować w zamienione dni
        if (
            !this.canWork(emp1, shift2.date) ||
            !this.canWork(emp2, shift1.date)
        )
            return;

        // WAŻNE: Sprawdź czy pracownicy nie mają już zmian w docelowe dni
        const emp1Dates = new Set(
            individual.shifts
                .filter((s, i) => s.employee_id === emp1.id && i !== idx1)
                .map((s) => s.date),
        );
        const emp2Dates = new Set(
            individual.shifts
                .filter((s, i) => s.employee_id === emp2.id && i !== idx2)
                .map((s) => s.date),
        );

        // emp1 dostaje shift2.date - sprawdź czy już ma zmianę w ten dzień
        if (emp1Dates.has(shift2.date)) return;
        // emp2 dostaje shift1.date - sprawdź czy już ma zmianę w ten dzień
        if (emp2Dates.has(shift1.date)) return;

        // Sprawdź hard constraints
        if (!this.checkSwapValid(individual.shifts, idx1, idx2)) return;

        // Wykonaj zamianę
        individual.shifts[idx1] = {
            ...shift1,
            employee_id: shift2.employee_id,
        };
        individual.shifts[idx2] = {
            ...shift2,
            employee_id: shift1.employee_id,
        };
    }

    private mutateMove(individual: Individual): void {
        if (individual.shifts.length === 0) return;

        const shiftIdx = Math.floor(Math.random() * individual.shifts.length);
        const shift = individual.shifts[shiftIdx];

        const emp = this.input.employees.find(
            (e) => e.id === shift.employee_id,
        );
        if (!emp) return;

        // Znajdź szablon dla tej zmiany (do sprawdzenia applicable_days)
        const template = this.input.templates.find(
            (t) => t.id === shift.template_id,
        );

        // Znajdź wolne dni
        const occupiedDates = new Set(
            individual.shifts
                .filter((s) => s.employee_id === shift.employee_id)
                .map((s) => s.date),
        );

        const freeDays = this.allWorkingDays.filter((day) => {
            if (occupiedDates.has(day)) return false;
            if (!this.canWork(emp, day)) return false;

            // ✅ Sprawdź czy szablon jest dostępny w ten dzień (applicable_days)
            if (
                template?.applicable_days &&
                template.applicable_days.length > 0
            ) {
                const dayKey = DAY_KEYS[getDayOfWeek(day)];
                if (!template.applicable_days.includes(dayKey as never))
                    return false;
            }

            return true;
        });

        if (freeDays.length === 0) return;

        const newDate = freeDays[Math.floor(Math.random() * freeDays.length)];

        // Sprawdź hard constraints
        const testShift = { ...shift, date: newDate };
        const otherShifts = individual.shifts.filter((_, i) => i !== shiftIdx);

        if (
            !checkHardConstraintsForShift(
                shift.employee_id,
                newDate,
                testShift,
                otherShifts,
            )
        )
            return;

        individual.shifts[shiftIdx] = testShift;
    }

    private mutateTemplate(individual: Individual): void {
        if (individual.shifts.length === 0) return;

        const shiftIdx = Math.floor(Math.random() * individual.shifts.length);
        const shift = individual.shifts[shiftIdx];

        const emp = this.input.employees.find(
            (e) => e.id === shift.employee_id,
        );
        if (!emp) return;

        // Oblicz dayKey dla dnia zmiany
        const shiftDayKey = DAY_KEYS[getDayOfWeek(shift.date)];

        // Znajdź dostępne szablony dla pracownika, które są też dostępne w ten dzień
        const availableTemplates = getAvailableTemplatesForEmployee(
            emp,
            this.input.templates,
            this.input.templateAssignmentsMap,
        ).filter((t) => {
            // ✅ Sprawdź czy szablon jest dostępny w ten dzień (applicable_days)
            if (t.applicable_days && t.applicable_days.length > 0) {
                if (!t.applicable_days.includes(shiftDayKey as never))
                    return false;
            }
            return true;
        });

        if (availableTemplates.length <= 1) return;

        // Wybierz inny szablon
        const otherTemplates = availableTemplates.filter(
            (t) => t.id !== shift.template_id,
        );
        if (otherTemplates.length === 0) return;

        const newTemplate =
            otherTemplates[Math.floor(Math.random() * otherTemplates.length)];

        individual.shifts[shiftIdx] = {
            ...shift,
            start_time: newTemplate.start_time.substring(0, 5),
            end_time: newTemplate.end_time.substring(0, 5),
            break_minutes: newTemplate.break_minutes ?? 0,
            template_id: newTemplate.id,
        };
    }

    // =========================================================================
    // POMOCNICZE
    // =========================================================================

    private clone(individual: Individual): Individual {
        return {
            shifts: individual.shifts.map((s) => ({ ...s })),
            fitness: individual.fitness,
        };
    }

    private groupByDay(
        shifts: GeneratedShift[],
    ): Map<string, GeneratedShift[]> {
        const map = new Map<string, GeneratedShift[]>();
        for (const shift of shifts) {
            if (!map.has(shift.date)) {
                map.set(shift.date, []);
            }
            map.get(shift.date)!.push(shift);
        }
        return map;
    }

    private fixConflicts(shifts: GeneratedShift[]): void {
        // Usuń duplikaty pracownik-dzień (zachowaj pierwszy)
        const seen = new Set<string>();
        const toRemove: number[] = [];

        for (let i = 0; i < shifts.length; i++) {
            const key = `${shifts[i].employee_id}-${shifts[i].date}`;
            if (seen.has(key)) {
                toRemove.push(i);
            } else {
                seen.add(key);
            }
        }

        // Usuń od końca żeby nie zaburzać indeksów
        for (let i = toRemove.length - 1; i >= 0; i--) {
            shifts.splice(toRemove[i], 1);
        }
    }

    private canWork(emp: EmployeeWithData, date: string): boolean {
        const dayOfWeek = new Date(date).getDay();
        const { holidays } = this.input;

        // Absencje
        if (emp.absences) {
            for (const absence of emp.absences) {
                if (date >= absence.start_date && date <= absence.end_date) {
                    return false;
                }
            }
        }

        // Preferencje (dane mogą być stringami lub liczbami w bazie)
        const unavailableDays =
            emp.preferences?.unavailable_days?.map((d: string | number) =>
                Number(d),
            ) || [];
        if (unavailableDays.includes(dayOfWeek)) {
            return false;
        }

        const isWeekend = dayOfWeek === 0 || dayOfWeek === 6;
        if (isWeekend && emp.preferences?.can_work_weekends === false) {
            return false;
        }

        const isHoliday = holidays.some((h) => h.date === date);
        if (isHoliday && emp.preferences?.can_work_holidays === false) {
            return false;
        }

        return true;
    }

    private checkSwapValid(
        shifts: GeneratedShift[],
        idx1: number,
        idx2: number,
    ): boolean {
        const shift1 = shifts[idx1];
        const shift2 = shifts[idx2];

        // Sprawdź hard constraints po zamianie
        const testShifts = shifts.map((s, i) => {
            if (i === idx1) return { ...s, employee_id: shift2.employee_id };
            if (i === idx2) return { ...s, employee_id: shift1.employee_id };
            return s;
        });

        // Sprawdź dla obu pracowników
        return (
            checkHardConstraintsForShift(
                shift2.employee_id,
                shift1.date,
                testShifts[idx1],
                testShifts.filter((_, i) => i !== idx1),
            ) &&
            checkHardConstraintsForShift(
                shift1.employee_id,
                shift2.date,
                testShifts[idx2],
                testShifts.filter((_, i) => i !== idx2),
            )
        );
    }
}

File: python-scheduler/scheduler/genetic_optimizer.py
"""
=============================================================================
ALGORYTM GENETYCZNY - OPTYMALIZACJA GRAFIKU
=============================================================================

Optymalizuje grafik metodą ewolucyjną.
Operatory genetyczne:
- Selekcja turniejowa
- Krzyżowanie jednopunktowe
- Mutacje: swap, move, change template
"""

import random
import logging
from typing import List, Tuple, Set
from datetime import datetime, timedelta
from copy import deepcopy

from .types import (
    SchedulerInput, 
    GeneratedShift, 
    Employee,
    ShiftTemplate,
    POLISH_LABOR_CODE
)
from .evaluator import evaluate_schedule, quick_fitness
from .validator import ScheduleValidator
from .utils import get_shift_time_type, get_day_of_week

logger = logging.getLogger(__name__)

# =============================================================================
# KONFIGURACJA
# =============================================================================

class GeneticConfig:
    """Konfiguracja algorytmu genetycznego"""
    def __init__(
        self,
        population_size: int = 30,
        generations: int = 100,
        mutation_rate: float = 0.2,
        crossover_rate: float = 0.7,
        elite_count: int = 2,
        tournament_size: int = 3,
        timeout_ms: int = 5000
    ):
        self.population_size = population_size
        self.generations = generations
        self.mutation_rate = mutation_rate
        self.crossover_rate = crossover_rate
        self.elite_count = elite_count
        self.tournament_size = tournament_size
        self.timeout_ms = timeout_ms

DEFAULT_CONFIG = GeneticConfig()

FAST_CONFIG = GeneticConfig(
    population_size=15,
    generations=30,
    mutation_rate=0.25,
    crossover_rate=0.6,
    elite_count=1,
    tournament_size=2,
    timeout_ms=2000
)

# =============================================================================
# OSOBNIK (ROZWIĄZANIE)
# =============================================================================

class Individual:
    """Osobnik w populacji - kompletny grafik"""
    
    def __init__(self, shifts: List[GeneratedShift]):
        self.shifts = shifts
        self.fitness = 0.0
    
    def clone(self) -> 'Individual':
        """Klonuje osobnika"""
        return Individual([
            GeneratedShift(
                employee_id=s.employee_id,
                date=s.date,
                start_time=s.start_time,
                end_time=s.end_time,
                break_minutes=s.break_minutes,
                template_id=s.template_id
            )
            for s in self.shifts
        ])

# =============================================================================
# GŁÓWNA KLASA OPTYMALIZERA
# =============================================================================

class GeneticOptimizer:
    """Optymalizator genetyczny dla grafików"""
    
    def __init__(
        self, 
        scheduler_input: SchedulerInput,
        config: dict = None
    ):
        self.input = scheduler_input
        
        # Konfiguracja
        if config:
            self.config = GeneticConfig(**config)
        else:
            self.config = DEFAULT_CONFIG
        
        # Dane pomocnicze
        self.all_working_days = sorted(
            scheduler_input.work_days + 
            scheduler_input.saturday_days + 
            scheduler_input.trading_sundays
        )
        
        self.weekend_days = set(
            scheduler_input.saturday_days + 
            scheduler_input.trading_sundays
        )
        
        self.validator = ScheduleValidator(scheduler_input)
        
        # Indeksy dla szybkiego dostępu
        self.employees_map = {emp.id: emp for emp in scheduler_input.employees}
        self.templates_map = {tmpl.id: tmpl for tmpl in scheduler_input.templates}
    
    # =========================================================================
    # GŁÓWNA METODA OPTYMALIZACJI
    # =========================================================================
    
    def optimize(self, initial_shifts: List[GeneratedShift]) -> List[GeneratedShift]:
        """
        Optymalizuje grafik algorytmem genetycznym
        
        Args:
            initial_shifts: Początkowy grafik do optymalizacji
            
        Returns:
            Zoptymalizowany grafik
        """
        start_time = datetime.now()
        
        logger.info("🧬 === GENETIC OPTIMIZER - START ===")
        logger.info(f"Populacja: {self.config.population_size}, "
                   f"Generacje: {self.config.generations}")
        
        # Fitness początkowy
        initial_fitness = quick_fitness(initial_shifts, self.input)
        logger.info(f"Fitness początkowy: {initial_fitness:.2f}")
        
        # Inicjalizacja populacji
        population = self._initialize_population(initial_shifts)
        
        best_individual = population[0]
        generations_without_improvement = 0
        
        # Główna pętla ewolucji
        for generation in range(self.config.generations):
            # Timeout
            elapsed_ms = (datetime.now() - start_time).total_seconds() * 1000
            if elapsed_ms > self.config.timeout_ms:
                logger.info(f"⏱️ Timeout po {generation} generacjach")
                break
            
            # Sortuj po fitness
            population.sort(key=lambda ind: ind.fitness, reverse=True)
            
            # Sprawdź postęp
            if population[0].fitness > best_individual.fitness:
                best_individual = population[0]
                generations_without_improvement = 0
                logger.info(f"Gen {generation}: Nowy najlepszy fitness = {best_individual.fitness:.2f}")
            else:
                generations_without_improvement += 1
            
            # Early stopping
            if generations_without_improvement > 20:
                logger.info(f"Brak poprawy przez 20 generacji, zatrzymuję")
                break
            
            # Nowa generacja
            new_population: List[Individual] = []
            
            # Elitaryzm - zachowaj najlepszych
            for i in range(self.config.elite_count):
                new_population.append(population[i].clone())
            
            # Generuj potomków
            while len(new_population) < self.config.population_size:
                parent1 = self._tournament_select(population)
                parent2 = self._tournament_select(population)
                
                # Krzyżowanie
                if random.random() < self.config.crossover_rate:
                    child = self._crossover(parent1, parent2)
                else:
                    child = parent1.clone()
                
                # Mutacja
                if random.random() < self.config.mutation_rate:
                    self._mutate(child)
                
                # Przelicz fitness
                child.fitness = quick_fitness(child.shifts, self.input)
                new_population.append(child)
            
            population = new_population
        
        # Najlepsze rozwiązanie
        population.sort(key=lambda ind: ind.fitness, reverse=True)
        best = population[0]
        
        improvement = ((best.fitness - initial_fitness) / initial_fitness * 100)
        
        logger.info(f"✅ Fitness końcowy: {best.fitness:.2f}")
        logger.info(f"📈 Poprawa: {improvement:.2f}%")
        logger.info("🧬 === GENETIC OPTIMIZER - END ===\n")
        
        return best.shifts
    
    # =========================================================================
    # INICJALIZACJA POPULACJI
    # =========================================================================
    
    def _initialize_population(self, base_shifts: List[GeneratedShift]) -> List[Individual]:
        """Inicjalizuje populację początkową"""
        population = []
        
        # Pierwszy osobnik - oryginalne rozwiązanie
        base_individual = Individual(base_shifts)
        base_individual.fitness = quick_fitness(base_shifts, self.input)
        population.append(base_individual)
        
        # Reszta populacji - wariacje
        for _ in range(self.config.population_size - 1):
            variant = base_individual.clone()
            
            # Losowe mutacje dla różnorodności
            num_mutations = random.randint(1, 5)
            for _ in range(num_mutations):
                self._mutate(variant)
            
            variant.fitness = quick_fitness(variant.shifts, self.input)
            population.append(variant)
        
        return population
    
    # =========================================================================
    # SELEKCJA
    # =========================================================================
    
    def _tournament_select(self, population: List[Individual]) -> Individual:
        """Selekcja turniejowa"""
        tournament = random.sample(population, self.config.tournament_size)
        return max(tournament, key=lambda ind: ind.fitness)
    
    # =========================================================================
    # KRZYŻOWANIE
    # =========================================================================
    
    def _crossover(self, parent1: Individual, parent2: Individual) -> Individual:
        """
        Krzyżowanie jednopunktowe
        Łączy zmiany z dwóch rodziców
        """
        # Grupuj zmiany po pracownikach
        p1_by_emp = {}
        p2_by_emp = {}
        
        for shift in parent1.shifts:
            if shift.employee_id not in p1_by_emp:
                p1_by_emp[shift.employee_id] = []
            p1_by_emp[shift.employee_id].append(shift)
        
        for shift in parent2.shifts:
            if shift.employee_id not in p2_by_emp:
                p2_by_emp[shift.employee_id] = []
            p2_by_emp[shift.employee_id].append(shift)
        
        # Losowy punkt cięcia
        all_employees = list(self.employees_map.keys())
        cutpoint = random.randint(1, len(all_employees) - 1)
        
        # Połącz
        child_shifts = []
        for i, emp_id in enumerate(all_employees):
            if i < cutpoint:
                child_shifts.extend(p1_by_emp.get(emp_id, []))
            else:
                child_shifts.extend(p2_by_emp.get(emp_id, []))
        
        return Individual(child_shifts)
    
    # =========================================================================
    # MUTACJE
    # =========================================================================
    
    def _mutate(self, individual: Individual):
        """
        Losowa mutacja osobnika
        Wybiera jeden z 3 typów mutacji
        """
        mutation_type = random.choice(['swap', 'move', 'change_template'])
        
        if mutation_type == 'swap':
            self._mutate_swap(individual)
        elif mutation_type == 'move':
            self._mutate_move(individual)
        else:
            self._mutate_change_template(individual)
    
    def _mutate_swap(self, individual: Individual):
        """Zamienia zmiany między dwoma pracownikami"""
        if len(individual.shifts) < 2:
            return
        
        # Wybierz dwie losowe zmiany
        shift1, shift2 = random.sample(individual.shifts, 2)
        
        # Zamień pracowników
        shift1.employee_id, shift2.employee_id = shift2.employee_id, shift1.employee_id
    
    def _mutate_move(self, individual: Individual):
        """Przenosi zmianę na inny dzień"""
        if not individual.shifts:
            return
        
        shift = random.choice(individual.shifts)
        
        # Wybierz inny dzień
        available_days = [d for d in self.all_working_days if d != shift.date]
        if not available_days:
            return
        
        new_date = random.choice(available_days)
        shift.date = new_date
    
    def _mutate_change_template(self, individual: Individual):
        """Zmienia szablon zmiany"""
        if not individual.shifts:
            return
        
        shift = random.choice(individual.shifts)
        employee = self.employees_map.get(shift.employee_id)
        
        if not employee or not employee.template_assignments:
            return
        
        # Wybierz inny szablon
        available_templates = [
            tmpl for tmpl in self.input.templates
            if tmpl.id in employee.template_assignments
        ]
        
        if not available_templates:
            return
        
        new_template = random.choice(available_templates)
        
        # Zaktualizuj zmianę
        shift.start_time = new_template.start_time
        shift.end_time = new_template.end_time
        shift.break_minutes = new_template.break_minutes
        shift.template_id = new_template.id
